runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/ProcessRemoteBundleOperation.java

```diff
@@ -100,6 +218,80 @@ public void finish() throws Exception {
       } catch (Exception e) {
         throw new RuntimeException("Failed to finish remote bundle", e);
       }
+
+      try {
+        // In Batch, the input watermark advances from -inf to +inf once all elements are processed.
+        // For timers, this means we only fire them at the end of the bundle.
+        fireTimers();
+      } catch (Exception e) {
+        throw new RuntimeException("Failed to finish remote bundle", e);
+      }
+    }
+  }
+
+  // Recursively fires all scheduled timers from the SDK Harness.
+  private void fireTimers() throws Exception {
+    DataflowExecutionContext.DataflowStepContext stepContext =
+        executionContext.getStepContext((DataflowOperationContext) this.context).namespacedToUser();
+
+    TimerInternals.TimerData timerData = stepContext.getNextFiredTimer(GlobalWindow.Coder.INSTANCE);
+    while (timerData != null) {
+      StateNamespaces.WindowNamespace windowNamespace =
+          (StateNamespaces.WindowNamespace) timerData.getNamespace();
+      BoundedWindow window = windowNamespace.getWindow();
+
+      WindowedValue<KV<Object, Timer>> timerValue =
+          WindowedValue.of(
+              KV.of(
+                  timerIdToKey.get(timerData.getTimerId()),
+                  Timer.of(timerData.getTimestamp(), timerIdToPayload.get(timerData.getTimerId()))),
+              timerData.getTimestamp(),
+              Collections.singleton(window),
+              PaneInfo.NO_FIRING);
+
+      String timerInputPCollection =
+          timerIdToTimerSpecMap.get(timerData.getTimerId()).inputCollectionId();
+
+      // TODO(BEAM-6274): check this for performance considerations.
+      // Timers are allowed to A) fire other timers, and B) fire themselves. To implement this,
+      // we send the timer to the SDK Harness, wait for any elements, then check if any additional
+      // timers were fired.
+      RemoteBundle timerRemoteBundle =
+          stageBundleFactory.getBundle(receiverFactory, stateRequestHandler, progressHandler);
+      timerRemoteBundle.getInputReceivers().get(timerInputPCollection).accept(timerValue);
+      timerRemoteBundle.close();
```

Comment:
Is this blocking? I'm not super familiar with the portability framework at this point.

Reply:
Yep, this is blocking. It needs to be blocking to wait for any timer elements to be scheduled by the SDK Harness.